### PR TITLE
perf(elasticsearch): Improve search and index performance

### PR DESCRIPTION
**Pull request by david-leifker**
Improve search and index performance when under high indexing

Added tests around when re-indexing occurs. For test speed, re-using elasticsearch instance across tests.

Added ability to override default index settings per index.

## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/indexbuilder/ESIndexBuilder.java [hunk 1]
```diff
@@ -53,6 +55,7 @@ public void buildIndex(String indexName, Map<String, Object> mappings, Map<Strin
     Map<String, Object> baseSettings = new HashMap<>(settings);
     baseSettings.put("number_of_shards", numShards);
     baseSettings.put("number_of_replicas", numReplicas);
+    baseSettings.put("refresh_interval", String.format("%ss", refreshIntervalSeconds));
```
Comment (RyanHolstien) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/indexbuilder/ESIndexBuilder.java:58]: Why are we doing a String.format here? Does Elastic need it to be a String?

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/indexbuilder/ESIndexBuilder.java [hunk 2]
```diff
@@ -43,6 +43,8 @@ public class ESIndexBuilder {
   private final int numReplicas;
   private final int numRetries;
 
+  private final int refreshIntervalSeconds;
+
   private static final List<String> SETTINGS_TO_COMPARE = ImmutableList.of("number_of_shards", "number_of_replicas");
```
Comment (RyanHolstien) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/indexbuilder/ESIndexBuilder.java:48]: It should also be added here to know if we need to update the index settings.

File: metadata-service/factories/src/main/resources/application.yml [hunk 3]
```diff
@@ -150,6 +150,7 @@ elasticsearch:
     numShards: ${ELASTICSEARCH_NUM_SHARDS_PER_INDEX:1}
     numReplicas: ${ELASTICSEARCH_NUM_REPLICAS_PER_INDEX:1}
     numRetries: ${ELASTICSEARCH_INDEX_BUILDER_NUM_RETRIES:3}
+    refreshIntervalSeconds: ${ELASTICSEARCH_INDEX_BUILDER_REFRESH_INTERVAL_SECONDS:30}
```
Comment (RyanHolstien) [inline metadata-service/factories/src/main/resources/application.yml:153]: I think we should keep 1 the default here, most deployments are not large enough to hit the write scaling issues that need the 30s setting.

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/indexbuilder/ESIndexBuilder.java (hunk 1 above)
Comment (david-leifker) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/indexbuilder/ESIndexBuilder.java:58]: Yes, requires a string with the unit 's' for seconds.

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/indexbuilder/ESIndexBuilder.java (hunk 2 above)
Comment (david-leifker) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/indexbuilder/ESIndexBuilder.java:48]: I see this information in the application.yml. Applying suggestion there.

Comment (RyanHolstien) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/indexbuilder/ESIndexBuilder.java:48]: Not sure what you mean here, I'm talking about the SETTINGS_TO_COMPARE immutable list which controls the start up check to see whether or not we should update the index settings during bootstrap.

Comment (david-leifker) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/indexbuilder/ESIndexBuilder.java:48]: Oh, I thought you were asking about the line that I added, basically asking for me to describe when to increase the interval seconds. Let me take a closer look at the SETTINGS_TO_COMPARE

Comment (david-leifker) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/indexbuilder/ESIndexBuilder.java]: Added the setting to this list & included a comment. Good catch here! Thanks!

**Review by RyanHolstien — APPROVED** [review]

**Comment by david-leifker** [comment]
There is a problem here. The index is going to be reindexed on a change to `refresh_interval` and this is not required. Need to write a new code path to apply settings without reindex when possible.

**Comment by david-leifker** [comment]
Something is not working with the smoke tests, investigating.

**Pull request closed without merging**